### PR TITLE
Include mGBA Printf to help debugging

### PR DESCRIPTION
**Pull request by AsparagusEduardo**
## Description
- Includes @SBird1337's implementation + Oldplayer's modification that allows encoded strings to be used in `MgbaPrintf`.

## **Discord contact info**
AsparagusEduardo#6051

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -109,6 +110,7 @@ void AgbMain()
     ClearDma3Requests();
     ResetBgs();
     SetDefaultFontsPointer();
+    MgbaOpen();
```
Comment (Jaizu) [inline src/main.c:113]: Suggestion: Add this under an ifdef/config option. Not everybody might want it enabled as default and there is no reason to force it on -afail-

**Comment by AsparagusEduardo** [comment]
Converted to draft until #2277 is merged

File: src/mgba_printf/mini_printf.c [hunk 2]
```diff
@@ -49,13 +49,90 @@
...
+#include "text.h"
 
 struct mini_buff 
 {
     char *buffer, *pbuffer;
     u32 buffer_len;
 };
 
+//This converts the number used to represent a character in the game code to a regular char (character).
+static u8 decodeChar(u8 input)
```
Comment (SBird1337) [inline src/mgba_printf/mini_printf.c:61]: This is incomplete, also referencing each alphanumerical character sounds a bit useless, also there is already an implementation here: <https://github.com/rh-hideout/pokeemerald-expansion/blob/c2d880fce9ba4fbe499da78aff0cb81ded65a257/src/mgba_printf/mgba.c#L46>

File: src/mgba_printf/mini_printf.c [hunk 3]
```diff
@@ -79,13 +156,43 @@ static s32 _puts(char *s, s32 len, void *buf)
     return b->pbuffer - p0;
 }
 
+static s32 _puts2(char *s, s32 len, void *buf)
```
Comment (SBird1337) [inline src/mgba_printf/mini_printf.c:159]: cryptic name, this is essentially `putsEncoded` or whatever

File: src/mgba_printf/mini_printf.c [hunk 4]
```diff
@@ -79,13 +156,43 @@ static s32 _puts(char *s, s32 len, void *buf)
...
+}
+
 static s32 mini_strlen(const char *s)
 {
 	s32 len = 0;
 	while (s[len] != '\0') len++;
 	return len;
 }
 
+static s32 mini_strlen2(const char *s)
```
Comment (SBird1337) [inline src/mgba_printf/mini_printf.c:189]: cryptic name, also essentially a copy of the library function `StringLength` that already exists in the game. It is debatable wether this should work as a standalone, but if not this should just use the string library function of the game.

File: src/mgba_printf/mini_printf.c [hunk 5]
```diff
@@ -167,13 +274,13 @@ s32 mini_vsnprintf(char *buffer, u32 buffer_len, const char *fmt, va_list va)
...
         buffer = NULL;
-    n = mini_vpprintf(_puts, (buffer != NULL) ? &b : NULL, fmt, va);
+    n = mini_vpprintf(_puts,_puts2, (buffer != NULL) ? &b : NULL, fmt, va);
     if (buffer == NULL)
         return n;
     return b.pbuffer - b.buffer;
 }
 
-s32 mini_vpprintf(s32 (*puts)(char* s, s32 len, void* buf), void* buf, const char *fmt, va_list va)
+s32 mini_vpprintf(s32 (*puts)(char* s, s32 len, void* buf),s32 (*puts2)(char* s, s32 len, void* buf), void* buf, const char *fmt, va_list va)
```
Comment (SBird1337) [inline src/mgba_printf/mini_printf.c:283]: The function essentially conforms to libc `vpprintf` which does not have a second function pointer argument. I think this should be reflected here, the alternate puts can just be called without a function pointer, essentially it is never called with another argument anyways.

File: src/mgba_printf/mini_printf.c [hunk 6]
```diff
@@ -275,10 +382,24 @@ s32 mini_vpprintf(s32 (*puts)(char* s, s32 len, void* buf), void* buf, const cha
...
+                    } else
+                    {
+                        len = puts2(ptr, len, buf);
+                    }
+                    break;
                 default:
                     len = 1;
                     len = puts(&ch, len, buf);
                     break;
+                    // MODIFIED: Use %S to convert to Ascii
```
Comment (SBird1337) [inline src/mgba_printf/mini_printf.c:401]: Comment is not really necessary, if the function shouod be documented, it should be done at the top

File: src/mgba_printf/mgba.c [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+static u32 MgbaConvertPStringToAsciiN(const char *src, char *dst, u32 n);
+static void MgbaPrintfBounded(s32 level, const char* ptr, ...);
+
+/*
+ * Converts a "PokeString" to ASCII. This function does not have any sanity checks for buffer lengths.
+ * Be careful not to overflow!
+ * Failed ASSERTIONS for buffers != NULL will halt the program!
+ */
+static u32 MgbaConvertPStringToAsciiN(const char *src, char *dst, u32 n)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:46]: If a pretty printing argument is provided we do not need an explicit function.

File: src/mgba_printf/mgba.c [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+            modifiedCode = '('; // opening parentheses
+        else if (*src == CHAR_RIGHT_PAREN)
+            modifiedCode = ')'; // closing parentheses
+        dst[i] = modifiedCode;
+    }
+    dst[i] = 0;
+    return i;
+}
+
+static void MgbaPrintfBounded(s32 level, const char* ptr, ...)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:98]: this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.

File: src/mgba_printf/mgba.c [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    va_list args;
+
+    level &= 0x7;
+    va_start(args, ptr);
+    mini_vsnprintf(REG_DEBUG_STRING, MGBA_REG_DEBUG_MAX, ptr, args);
+    va_end(args);
+    *REG_DEBUG_FLAGS = level | 0x100;
+}
+
+void MgbaPrintf(s32 level, const char* ptr, ...)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:109]: this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.

File: src/mgba_printf/mgba.c [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+                break;
+        }
+        offset += i;
+        *REG_DEBUG_FLAGS = level | 0x100;
+    } while ((i == MGBA_REG_DEBUG_MAX) && (buffer[offset] != '\0'));
+
+    Free(buffer);
+}
+
+bool8 MgbaOpen(void)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:142]: this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.

File: src/mgba_printf/mgba.c [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+{
+    *REG_DEBUG_ENABLE = 0xC0DE;
+    return *REG_DEBUG_ENABLE == 0x1DEA;
+}
+
+void MgbaClose(void)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:148]: this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.

File: src/mgba_printf/mgba.c [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+{
+    *REG_DEBUG_ENABLE = 0;
+}
+
+void MgbaAssert(const char *pFile, s32 nLine, const char *pExpression, bool32 nStopProgram)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:153]: this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.

File: src/mgba_printf/mgba.c [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+    else
+    {
+        MgbaPrintfBounded(MGBA_LOG_WARN, "WARING FILE=[%s] LINE=[%d]  EXP=[%s]", pFile, nLine, pExpression);
+    }
+}
+
+/*
+ * Prints a "PokeString" to the MGBA LOG specified with `level`.
+ */
+void MgbaPrintEncoded(s32 level, const char *encodedString)
```
Comment (SBird1337) [inline src/mgba_printf/mgba.c:169]: If a pretty printing argument is provided we do not need this function

**Review by SBird1337 — CHANGES_REQUESTED** [review]
A few things I noticed. Keep in mind that this is for printing messages on an mgba log. If you try to write to the virtual registers on any other platforms it will not work as expected or crash. There's also quite a lot of either incomplete or duplicate code in this PR.

**Pull request closed without merging**